MovingBlocks/Terasology
# [NFM] StateManager for DAG

### Contains

Add StateManager, StateManagerTest, State, AbstractNode, Node, RenderingGraph and so on.

#### Structure:

In _dag-statemanager_ branch, structure of _dag_ package is:

* dag
	* nodes**
		* AmbientOcclusionPassesNode
		* ...
		* BackdropNode*
		* LightGeometryNode*
		* SkyBandsNode*
		* ChunksRefractiveReflectiveNode*
		* ChunksAlphaRejectNode*
		* ChunksOpaqueNode*
		* FirstPersonViewNode*
		* ObjectsOpaqueNode*
		* OverlaysNode*
		* ...
		* WorldReflectionNode
	* states
		* AbstractState
		* State
		* StateType
		* StateTypeImpl
		* StateValue
		* Wireframe
	* AbstractNode
	* Node
	* NotAnalyzedException
	* RenderingGraph
	* StateManager

\*: _node desires Wireframe state by_ `addDesiredState(StateTypeImpl.WIREFRAME, StateValue.ENABLED)`.
\**: _important to note that these nodes are very outdated when it's compared to ones that are already merged and accepted to `develop` branch._

#### Definitions:

**RenderingGraph:** Stores all nodes and a _StateManager_ instance, generates (will be) topological order of the graph. RenderingGraph can not be "processed", if it is not analyzed by the _StateManager_ instance. If `process` is invoked before invocation of `analyze()`, RenderingGraph throws a custom *NotAnalyzedException*.

**StateManager:** Gets topologically sorted rendering pipeline(not sorted now, since a basic insert-ordered list is used), prepares state changes by `findStateChanges(Collection<Node> nodes)` and executes these changes by `prepareFor(Node node)` method.

For demonstrating the need of **State**, **StateType**, **StateTypeImpl** and **StateValue** please check **StateManagerTest** class.

A snippet for quick reviewing:

```java
@Test
public void testOneStateSixNodes2() {
        StateManager stateManager = new StateManager();
        DummyState dummyStateA = new DummyState();
        stateManager.addState(DummyStateType.A, dummyStateA);
        DummyNode node1 = new DummyNode("node1");
        DummyNode node2 = new DummyNode("node2");
        DummyNode node3 = instantiateDummyNodeWithState("node3", DummyStateType.A, StateValue.ENABLED);
        DummyNode node4 = instantiateDummyNodeWithState("node4", DummyStateType.A, StateValue.ENABLED);
        DummyNode node5 = new DummyNode("node5");
        DummyNode node6 = new DummyNode("node6");

        Collection<Node> dummyPipeline = createDummyPipeline(node1, node2, node3, node4, node5, node6);
        stateManager.findStateChanges(dummyPipeline);

        // run pipeline
        for (Node node : dummyPipeline) {
            stateManager.prepareFor(node); 
            node.process() // doesn't have any effect on the test
        }

        assertEquals("10", dummyStateA.getHistory());
    }

```
 



#### _(Unofficial)_ Abbreviations:

**NFM**: _Not for Merge_
**WIP**: _Work in Progress_
**RFR**: _Ready for Review_
**RFM**: _Ready for Merge_

_Would be great to have @cervator's opinion about these, plus making them official perhaps?_

### How to test

* Run _StateManagerTest_ for testing _StateManager_.
* Play the game with different video settings and observe any irregularities.  

### Outstanding before merging


- [ ] Naming, patterns and architecture?, from @manu3d:

> 
* [ ] a **RenderPipelineGenerator** that takes the current render graph in input and generates the render pipeline 
* [ ] once generated the render pipeline should be printed out in the log, so we can inspect it
a **RenderPipelineOptimizer** that takes an existing pipeline and optimizes it by removing redundant state changes
* [ ] once again, we should get a print out to inspect the outcome of the optimization process 
* [ ] the rendering occurring by processing the rendering pipeline and generating images that are identical to pre-dag changes.
